KAFKA-9274: Remove `retries` from InternalTopicManager

mjsax opened the pull request:
- part of KIP-572
 - replace `retries` in InternalTopicManager with infinite retires plus
   a new timeout, based on consumer config MAX_POLL_INTERVAL_MS
- if the new timeout hits, we don't throw `StreamsException` any longer (as we did when exceeding retries), but send `INCOMPLETE_SOURCE_TOPIC_METADATA` error code to let all instances shut down

Third PR for KIP-572 (cf #8864 and #9047)

Call for review @vvcephei

mjsax committed:
KAFKA-9274: Remove retries from InternalTopicManager
 - part of KIP-572
 - replace `retries` in InternalTopicManager with infinite retires plus
   a new timeout, based on consumer config MAX_POLL_INTERVAL_MS

mjsax committed:
fix tests

mjsax commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1249,11 +1248,7 @@ public void cleanUp() {
-        } catch (final TimeoutException e) {
-            throw new StreamsException("Timed out obtaining end offsets from kafka", e);
-        }
+        allEndOffsets = fetchEndOffsets(allPartitions, adminClient);
```
Minor side fix: `fetchEndOffset` can never throw `TimeoutException` because it catches all `RuntimeException` and convert them into `StreamsException` already

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -46,33 +49,38 @@
+
+    private final Map<String, String> defaultTopicConfigs = new HashMap<>();
 
-    @SuppressWarnings("deprecation") // TODO: remove in follow up PR when `RETRIES` is removed
```
first TODO removed

mjsax commented on streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java:
```diff
@@ -159,7 +159,6 @@ public void testGetGroupInstanceIdConfigs() {
         assertNull(returnedProps.get(ConsumerConfig.GROUP_INSTANCE_ID_CONFIG));
     }
 
-    @SuppressWarnings("deprecation") // TODO revisit in follow up PR
```
Second TODO removed

mjsax commented on streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java:
```diff
@@ -186,25 +183,9 @@ public void consumerConfigMustContainStreamPartitionAssignorConfig() {
         assertEquals(100, returnedProps.get(StreamsConfig.topicPrefix(TopicConfig.SEGMENT_BYTES_CONFIG)));
     }
 
-    @SuppressWarnings("deprecation") // TODO revisit in follow up PR
```
Third TODO removed -- as we don't pass `retires` and `retry.backoff.ms` via admin config any longer, this test is not needed any more

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java:
```diff
@@ -71,21 +73,20 @@
     private MockAdminClient mockAdminClient;
     private InternalTopicManager internalTopicManager;
 
-    @SuppressWarnings("deprecation") // TODO revisit in follow up PR
```
Fourth TODO removed

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -154,27 +164,34 @@ public InternalTopicManager(final Admin adminClient, final StreamsConfig streams
+                    final String timeoutError = String.format("Could not create topics within %d milliseconds. " +
+                        "This can happen if the Kafka cluster is temporary not available.", retryTimeoutMs);
+                    log.error(timeoutError);
+                    // TODO: should we throw a different exception instead and catch it, to return a `INCOMPLETE_SOURCE_TOPIC_METADATA` error code
```
@vvcephei this is an open question

\cc @guozhangwang

guozhangwang replied:
I'm thinking that moving forward we should try to not create internal topics during rebalance but try pre-create in starting, but for now assuming this is still the case I think letting the whole application to die is fine --- i.e. treat it the same as source topics. Hence I'm leaning towards encoding INCOMPLETE_SOURCE_TOPIC_METADATA to shutdown the whole app, across all clients.

mjsax replied:
Ack

mjsax commented:
Triggered system tests: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4071/

dajac commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -96,13 +104,15 @@ public InternalTopicManager(final Admin adminClient, final StreamsConfig streams
 
-        int remainingRetries = retries;
+        long currentWallClockMs = time.milliseconds();
+        final long deadlineMs = currentWallClockMs + retryTimeoutMs;
```
If we want to guarantee that the `deadlineMs` is respected, I think that we must set the timeout of the AdminClient's call accordingly: `CreateTopicsOptions.timeoutMs`. With the default, I think that the call could be longer than half of `MAX_POLL_INTERVAL_MS_CONFIG`.

mjsax replied:
Good question. Default `max.poll.interval.ms` is 5 minutes (ie, the deadline is set to 2.5 minutes by default) while default `api.default.timeout.ms` is 1 minutes? Thus we might be ok?

dajac replied:
That's right. I misread the default value of `max.poll.interval.ms`, too many zeros for my eyes ;). The default works fine then. Do we want to protect ourselves if the user changes the default? Or shall we just call out that `api.default.timeout.ms` should be lower than `max.poll.interval.ms` somewhere?

mjsax replied:
I am happy to add a check in `StreamsConfig` and either throw or log a WARN depending how strict we want to be.

dajac replied:
Thinking a bit more about this, with the default, you may end up not honouring the deadline. `createTopics` can take up to 1m so if you invoke one when less than 1m is reaming before the deadline, you may not honour the deadline. It may not be that important though.

If we want to strictly enforce it, we could calculate the maximum timeout for each call, something like deadline - now, and set it with `CreateTopicsOptions.timeoutMs`.

mjsax commented:
Only the `StreamsStandbyTask.test_standby_tasks_rebalance` system test failed and it's know to be buggy.

mjsax commented:
Jenkins failed on know flaky tests only.

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java:
```diff
@@ -71,21 +73,20 @@
             put(StreamsConfig.REPLICATION_FACTOR_CONFIG, 1);
             put(StreamsConfig.producerPrefix(ProducerConfig.BATCH_SIZE_CONFIG), 16384);
-            put(StreamsConfig.adminClientPrefix(StreamsConfig.RETRIES_CONFIG), numRetries);
+            put(StreamsConfig.consumerPrefix(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG), 100);
```
Why do we reduce max poll interval?

mjsax replied:
In this PR, we change the "deadline" in the group leader to create/verify all internal topics from "counting retries" to a timeout of `max.poll.interval.ms / 2` and we reduce the default of 5 minutes to speed up this test.

cf https://github.com/apache/kafka/pull/9060/files#diff-d3963e433c59b08688bb4481faa20e97R79

abbccdda reviewed (commented):
Overall LGTM

abbccdda commented:
Also, do we have new unit test coverage for the changes?

mjsax committed:
Github comments

mjsax commented:
Updated. Call for review @abbccdda 

> Also, do we have new unit test coverage for the changes?

We did not really change much, only switching from `retries` to a "timeout" and thus existing unit tests (that I updated accordingly, eg, different configs and/or different exception type) should be sufficient?

The only other thing is the new error code and I just added 2 unit tests for this case.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -373,8 +373,15 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
-            assignTasksToClients(fullMetadata, allSourceTopics, topicGroups, clientMetadataMap, partitionsForTask, statefulTasks);
+        final boolean probingRebalanceNeeded;
+        try {
+            probingRebalanceNeeded = assignTasksToClients(fullMetadata, allSourceTopics, topicGroups, clientMetadataMap, partitionsForTask, statefulTasks);
```
Is this part of the initiative to throw a different exception? Could we update the summary of this PR?

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -477,7 +484,7 @@ private boolean checkMetadataVersions(final int minReceivedMetadataVersion,
      */
     private Map<TopicPartition, PartitionInfo> prepareRepartitionTopics(final Map<Integer, TopicsInfo> topicGroups,
-                                                                           final Cluster metadata) {
+                                                                           final Cluster metadata) throws TaskAssignmentException {
```
Could we update the meta comment to explain when a `TaskAssignmentException` is thrown?

mjsax replied:
`TaskAssignmentException` is not a checked exception and it's just a curtesy declaration... (I can also remove `throws TaskAssignmentException` if you prefer).

Don't see any need to document anything further -- the code makes it clear (in fact, this method does not even throw the exception itself, but it just bubbles up from `internalTopicManager.makeReady` and the code documents itself:
https://github.com/apache/kafka/pull/9060/files#diff-d3963e433c59b08688bb4481faa20e97R179-R184

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -91,18 +100,20 @@ public InternalTopicManager(final Admin adminClient, final StreamsConfig streams
      * @return the set of topics which had to be newly created
      */
-    public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
+    public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) throws TaskAssignmentException {
```
Comments for thrown exception

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java:
```diff
@@ -383,18 +405,25 @@ public void shouldThrowExceptionWhenKeepsTopicLeaderNotAvailable() {
+        assertThat(
+            exception.getMessage(),
+            equalTo("Could not create topics within 50 milliseconds." +
+                " This can happen if the Kafka cluster is temporary not available.")
```
s/temporary/temporarily

mjsax committed:
Github comments

mjsax commented:
@abbccdda I updated the PR. I also realized, that it might be better to throw a `TimeoutException` instead of a `TaskAssignmentException` if we hit the new timeout. Updated the code and tests accordingly.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -373,8 +373,15 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
+        try {
+            probingRebalanceNeeded = assignTasksToClients(fullMetadata, allSourceTopics, topicGroups, clientMetadataMap, partitionsForTask, statefulTasks);
+        } catch (final TaskAssignmentException | TimeoutException e) {
+            return new GroupAssignment(
```
Since we could throw different exceptions here, would be good to add a log to indicate which type of exception is thrown.

mjsax replied:
There are already corresponding log.error statement before those exceptions are thrown. No need to double log IMHO?

abbccdda replied:
Cool, sg.

abbccdda reviewed (commented):
Just one minor comment.

mjsax committed:
Rebased

abbccdda reviewed (approved):
LGTM

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -45,33 +49,38 @@
+        // need to add mandatory configs; otherwise `QuietConsumerConfig` throws
+        consumerConfig.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class);
+        consumerConfig.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class);
+        retryTimeoutMs = new QuietConsumerConfig(consumerConfig).getInt(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG) / 2L;
```
Hey @mjsax , am I reading this PR correctly? Do we now only allow a single member to retry topic creation/validation for up to half of the poll interval, after which we shut down the entire application? That sounds like the opposite of resiliency...what if the brokers are temporarily unavailable? Before this we would just let the single thread die, and the internal topic creation/validation would be retried on the subsequent rebalance. That wasn't ideal, but given the upcoming work to allow reviving/recreating a death thread, that seems to be preferable to permanently ending the application? 

Sorry if I'm misreading this, was just going over all the PRs in the last month or so to produce a diff+summary of the important ones, and want to make sure I actually understand all the changes we've made

ableegoldman replied:
Apologies if this was touched on in the KIP, it's been a while and the discussion thread was quite long so I may have missed something there

mjsax replied:
Note that the previous default was "zero retries" and thus the new default is more resilient with a 5 minute default max.poll.interval. -- But yes, we shutdown the whole app for this case now as proposed by @guozhangwang (IIRC).

guozhangwang replied:
Today since we do not have ways to partially create tasks we'd have to create all topics to make sure all tasks are "complete" within each rebalance, if we cannot successfully create the topics within the poll.interval (i.e. we'd need to complete that rebalance with the poll.interval, and I guess halving it is to be more conservative), then killing that thread is not very useful anyways since we cannot proceed with the initializable tasks anyways.

That being said, with the upcoming work I'd agree that just shutdown the thread and allow users to optionally retry rebalance with new threads would be preferrable.

guozhangwang replied:
cc @cadonna @wcarlson5 to bring to your attention.

ableegoldman replied:
Yeah I think we should remove the shutdown error code in case of TimeoutException during internal topic validation before 2.7. I'll create a ticket so we don't lose track -- I think even just letting it kill the one thread is better than killing all of them

mjsax replied:
SGTM.
